### ISSUE TITLE
Out of bounds error of wight[ignore_index] in torch.nn.fucntional.cross_entropy() only when using weight, label_smoothing, and reduction='mean'.

### ISSUE THREAD

**Issue by indiaki1o** [pytorch/pytorch#91383]
### 🐛 Describe the bug

An IndexError occurs in `torch.nn.functional.cross_entropy(Input, Target, ...)` when `label_smoothing>0`, `weight` is given, `reduction='mean'`, and `ignore_index` is other than $[0, N_\mathrm{class})$ in `Target` tonsor.
It seems `weight[Target]` was referred to without ignoring data of `ignore_index` in `Target`, then causing the out-of-bounds error in `torch._C._nn.cross_entropy_loss`.

FYI: It doesn't occur when `label_smoothing=0`, `reduction='sum'` or `'none'`, or weigh is None.

```{python}
import numpy as np
import sys
print("python={}".format(sys.version_info) )
print("torch.__version__={}".format(torch.__version__))
# python=sys.version_info(major=3, minor=7, micro=11, releaselevel='final', serial=0)
# torch.__version__=1.10.0

# Input shapes (N=3, C=3)
_ones = torch.ones(N,1)
_zeros = torch.zeros(N,1)
X = torch.cat([_ones, _zeros, _zeros], dim=1) # respond all sample as class 0
# torch.Size([3, 3])

# Target shapes (N=3,) with ignore_index=-100
Y = torch.tensor([0,1,-100]).type(torch.LongTensor)
print(Y) # tensor([   0,    1, -100])

weight=torch.ones(3) # for three classes
_ = torch.nn.functional.cross_entropy(X,Y, weight=weight, ignore_index=-100, label_smoothing=0,   reduction='mean') # label_smoothing=0
_ = torch.nn.functional.cross_entropy(X,Y, weight=None,   ignore_index=-100, label_smoothing=0.1, reduction='mean') # wieght=None
_ = torch.nn.functional.cross_entropy(X,Y, weight=weight, ignore_index=-100, label_smoothing=0.1, reduction='sum') # reduction='sum'
_ = torch.nn.functional.cross_entropy(X,Y, weight=weight, ignore_index=-100, label_smoothing=0.1, reduction='none') # reduction='none'
_ = torch.nn.functional.cross_entropy(X,Y, weight=weight, ignore_index=-100, label_smoothing=0.1, reduction='mean') # error is only here
```

```
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
/tmp/ipykernel_2483606/2413739921.py in <module>
      4 _ = torch.nn.functional.cross_entropy(X,Y, weight=weight, ignore_index=-100, label_smoothing=0.1, reduction='sum')
      5 _ = torch.nn.functional.cross_entropy(X,Y, weight=weight, ignore_index=-100, label_smoothing=0.1, reduction='none')
----> 6 _ = torch.nn.functional.cross_entropy(X,Y, weight=weight, ignore_index=-100, label_smoothing=0.1, reduction='mean') # error is here

/Anaconda/envs/py37/lib/python3.7/site-packages/torch/nn/functional.py in cross_entropy(input, target, weight, size_average, ignore_index, reduce, reduction, label_smoothing)
   2844     if size_average is not None or reduce is not None:
   2845         reduction = _Reduction.legacy_get_string(size_average, reduce)
-> 2846     return torch._C._nn.cross_entropy_loss(input, target, weight, _Reduction.get_enum(reduction), ignore_index, label_smoothing)
   2847 
   2848 

RuntimeError: index -100 is out of bounds for dimension 0 with size 3
```

### Versions

Collecting environment information...
PyTorch version: 1.10.0
Is debug build: False
CUDA used to build PyTorch: 11.3
ROCM used to build PyTorch: N/A

OS: Ubuntu 18.04.6 LTS (x86_64)
GCC version: (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0
Clang version: Could not collect
CMake version: Could not collect
Libc version: glibc-2.17

Python version: 3.7.11 (default, Jul 27 2021, 14:32:16)  [GCC 7.5.0] (64-bit runtime)
Python platform: Linux-5.15.0-53-generic-x86_64-with-debian-buster-sid
Is CUDA available: True
CUDA runtime version: 11.3.109
CUDA_MODULE_LOADING set to: 
GPU models and configuration: GPU 0: NVIDIA GeForce RTX 3080
Nvidia driver version: 515.43.04
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

Versions of relevant libraries:
[pip3] efficientnet-pytorch==0.6.3
[pip3] mypy-extensions==0.4.3
[pip3] numpy==1.21.2
[pip3] numpydoc==1.1.0
[pip3] segmentation-models-pytorch==0.2.1
[pip3] torch==1.10.0
[pip3] torchaudio==0.10.0
[pip3] torchinfo==1.7.0
[pip3] torchvision==0.11.1
[pip3] torchviz==0.0.2
[conda] blas                      1.0                         mkl  
[conda] cudatoolkit               11.3.1               h2bc3f7f_2  
[conda] efficientnet-pytorch      0.6.3              pyh9f0ad1d_0    conda-forge
[conda] mkl                       2021.4.0           h06a4308_640  
[conda] mkl-service               2.4.0            py37h7f8727e_0  
[conda] mkl_fft                   1.3.1            py37hd3c417c_0  
[conda] mkl_random                1.2.2            py37h51133e4_0  
[conda] numpy                     1.21.2           py37h20f2e39_0  
[conda] numpy-base                1.21.2           py37h79a1101_0  
[conda] numpydoc                  1.1.0              pyhd3eb1b0_1  
[conda] pytorch                   1.10.0          py3.7_cuda11.3_cudnn8.2.0_0    pytorch
[conda] pytorch-mutex             1.0                        cuda    pytorch
[conda] segmentation-models-pytorch 0.2.1              pyhd8ed1ab_0    conda-forge
[conda] torchaudio                0.10.0               py37_cu113    pytorch
[conda] torchinfo                 1.7.0              pyhd8ed1ab_0    conda-forge
[conda] torchvision               0.11.1               py37_cu113    pytorch
[conda] torchviz                  0.0.2                    pypi_0    pypi

cc @albanD @mruberry @jbschlosser @walterddr @kshitij12345 @saketh-are

**Comment by ZailiWang** [comment]
Hi, I tried with  PyTorch1.12 & 1.13 env at hand and did not re-produce the error. Would you update to latest version?

**Comment by albanD** [comment]
I can't reproduce either on latest PyTorch, this has most likely been fixed since 1.10.
Feel free to re-open if you still see this error on latest version.
